pkg/scheduler/internal/cache/cache_test.go

```diff
@@ -1117,6 +1126,292 @@ func TestNodeOperators(t *testing.T) {
+				{operator: "AddNode", index: 2},
+				{operator: "AddNode", index: 5},
+				{operator: "AddNode", index: 6},
+				{operator: "Snapshot"},
+				{operator: "RemoveNode", index: 2},
+				{operator: "RemoveNode", index: 5},
+				{operator: "RemoveNode", index: 6},
+				{operator: "Snapshot"},
+				{operator: "AddNode", index: 7},
+				{operator: "AddNode", index: 9},
+				{operator: "Snapshot"},
+			},
+			expected: []*v1.Node{nodes[9], nodes[7]},
+		},
+		{
+			name: "Update nodes in particular order",
+			operations: []operation{
+				{operator: "AddNode", index: 8},
+				{operator: "UpdateNode", index: 2},
+				{operator: "UpdateNode", index: 8},
+				{operator: "Snapshot"},
+				{operator: "AddNode", index: 1},
+				{operator: "Snapshot"},
+			},
+			expected: []*v1.Node{nodes[1], nodes[8], nodes[2]},
+		},
+		{
+			name: "Add some nodes and some pods",
+			operations: []operation{
+				{operator: "AddNode", index: 0},
+				{operator: "AddNode", index: 2},
+				{operator: "AddNode", index: 8},
+				{operator: "Snapshot"},
+				{operator: "AddPod", index: 8},
+				{operator: "AddPod", index: 2},
+				{operator: "Snapshot"},
+			},
+			expected: []*v1.Node{nodes[2], nodes[8], nodes[0]},
+		},
+		{
+			name: "Updating a pod moves its node to the head",
+			operations: []operation{
+				{operator: "AddNode", index: 0},
+				{operator: "AddPod", index: 0},
+				{operator: "AddNode", index: 2},
+				{operator: "AddNode", index: 4},
+				{operator: "UpdatePod", index: 0},
+				{operator: "Snapshot"},
+			},
+			expected: []*v1.Node{nodes[0], nodes[4], nodes[2]},
+		},
+		{
+			name: "Remove pod from non-existing node",
+			operations: []operation{
+				{operator: "AddNode", index: 0},
+				{operator: "AddPod", index: 0},
+				{operator: "AddNode", index: 2},
+				{operator: "Snapshot"},
+				{operator: "RemovePod", index: 3},
+				{operator: "Snapshot"},
+			},
+			expected: []*v1.Node{nodes[2], nodes[0]},
+		},
+	}
+
+	for _, test := range tests {
+		cache := newSchedulerCache(time.Second, time.Second, nil)
+		snapshot := NewNodeInfoSnapshot()
+		snapshotTaken := false
+
+		for _, op := range test.operations {
+			switch op.operator {
+			case "AddNode":
+				cache.AddNode(nodes[op.index])
+			case "RemoveNode":
+				cache.RemoveNode(nodes[op.index])
+			case "UpdateNode":
+				cache.UpdateNode(nodes[op.index], updatedNodes[op.index])
+			case "AddPod":
+				cache.AddPod(pods[op.index])
+			case "RemovePod":
+				cache.RemovePod(pods[op.index])
+			case "UpdatePod":
+				cache.UpdatePod(pods[op.index], updatedPods[op.index])
+			case "Snapshot":
+				cache.SnapshotNodeInfo(&snapshot)
+				snapshotTaken = true
+			default:
+				t.Errorf("Invalid operator: %v", op.operator)
+			}
+		}
+
+		if len(test.expected) != len(cache.nodes) {
+			t.Errorf("Test: %v, unexpected number of nodes. Expected: %v, got: %v", test.name, len(test.expected), len(cache.nodes))
+		}
+		var i int
+		for node := cache.headNode; node != nil; node = node.next {
```

Comment:
It may be obvious from reading the code, but could we add a comment to this and the following for loop that one is checking internal cache state and the other is checking snapshot data?

Reply:
done